File: src/main/java/io/gittul/domain/post/dto/PostFeedResponse.java
package io.gittul.domain.post.dto;

import io.gittul.domain.comment.dto.CommentResponse;
import io.gittul.domain.post.entity.Post;
import io.gittul.domain.tag.entity.Tag;
import io.gittul.domain.user.dto.UserProfileResponse;
import io.gittul.domain.user.entity.User;

import java.util.List;

public record PostFeedResponse(
        UserProfileResponse user,
        String title,
        String image,
        String description,
        String createdAt,
        String updatedAt,
        long id,
        int starCount,
        int forkCount,
        int viewCount,
        int likeCount,
        int commentCount,
        CommentResponse bestComment,
        boolean isLiked,
        boolean isBookmarked,
        List<String> tags
) {

    // Todo. 명칭 수정 post -> thread
    public static PostFeedResponse ofAndTo(Post thread, User requestingUser) {
        return createPostFeedResponse(
                thread,
                thread.isLikedBy(requestingUser),
                thread.isBookmarkedBy(requestingUser)
        );
    }

    public static PostFeedResponse ofNew(Post thread) {
        return createPostFeedResponse(thread, false, false);
    }

    private static PostFeedResponse createPostFeedResponse(Post thread,
                                                           boolean isLiked,
                                                           boolean isBookmarked) {

        CommentResponse bestCommentResponse = thread.getBestComment() != null
                ? CommentResponse.ofAndTo(thread.getBestComment(), thread.getUser())
                : null;

        return new PostFeedResponse(
                UserProfileResponse.of(thread.getUser()),
                thread.getTitle(),
                thread.getImageUrl(),
                thread.getContent(),
                thread.getCreatedAt().toString(),
                thread.getUpdatedAt().toString(),
                thread.getPostId(),
                thread.getRepository() != null ? thread.getRepository().getStarCount() : 0,
                thread.getRepository() != null ? thread.getRepository().getForkCount() : 0, // Todo. 가독성 개선
                thread.getViewCount(),
                thread.getLikeCount(),
                thread.getCommentCount(),
                bestCommentResponse,
                isLiked,
                isBookmarked,
                thread.getTags().stream().map(Tag::getTagName).toList()
        );
    }
}

File: app/src/main/kotlin/io/gittul/app/domain/post/dto/PostFeedResponse.kt
package io.gittul.app.domain.post.dto

import io.gittul.app.domain.comment.dto.CommentResponse
import io.gittul.app.domain.user.dto.UserProfileResponse
import io.gittul.core.domain.post.entity.Post
import io.gittul.core.domain.user.entity.User

data class PostFeedResponse(
    val user: UserProfileResponse,
    val title: String,
    val image: String?,
    val description: String,
    val createdAt: String,
    val updatedAt: String,
    val id: Long,
    val starCount: Int,
    val forkCount: Int,
    val viewCount: Int,
    val likeCount: Int,
    val commentCount: Int,
    val bestComment: CommentResponse?,
    val isLiked: Boolean,
    val isBookmarked: Boolean,
    val tags: List<String>
) {
    companion object {
        // Todo. 명칭 수정 post -> thread
        fun ofAndTo(thread: Post, requestingUser: User?): PostFeedResponse {
            return createPostFeedResponse(
                thread,
                thread.isLikedBy(requestingUser),
                thread.isBookmarkedBy(requestingUser)
            )
        }

        fun ofNew(thread: Post): PostFeedResponse {
            return createPostFeedResponse(thread, false, false)
        }

        private fun createPostFeedResponse(
            thread: Post,
            isLiked: Boolean,
            isBookmarked: Boolean
        ): PostFeedResponse {
            val bestCommentResponse: CommentResponse? = thread.bestComment?.let {
                CommentResponse.ofAndTo(it, thread.user)
            }

            return PostFeedResponse(
                user = UserProfileResponse.of(thread.user),
                title = thread.title,
                image = thread.imageUrl,
                description = thread.content,
                createdAt = thread.createdAt.toString(),
                updatedAt = thread.updatedAt.toString(),
                id = thread.postId,
                starCount = thread.repository?.starCount ?: 0,
                forkCount = thread.repository?.forkCount ?: 0,  // Todo. 가독성 개선
                viewCount = thread.viewCount,
                likeCount = thread.likeCount,
                commentCount = thread.commentCount,
                bestComment = bestCommentResponse,
                isLiked = isLiked,
                isBookmarked = isBookmarked,
                tags = thread.tags.map { it.tagName }
            )
        }
    }
}
